From aced54e554c6f757317c4e8b9126d9254531d9eb Mon Sep 17 00:00:00 2001
From: Matthew Russell <matthew.g.russell@gmail.com>
Date: Thu, 22 Jun 2023 14:43:42 +0000
Subject: [PATCH] Allow service_discovery to continue without random_device

Allow service_discovery to continue even if the random_device isn't yet
available.  This is unlikely to occur on linux systems, but not
impossible.  It is more likely on other systems such as QNX.  In the
event that the random device isn't available, an error message is
produced and the middle of the min/max for the delay is used.

Without this change, if the random device is missing, the following
error is thrown:

   terminate called after throwing an instance of 'std::__1::system_error'
   what():  random_device failed to open /dev/urandom: No such file or directory
   Abort (core dumped)
---
 .../src/service_discovery_impl.cpp            | 30 +++++++++++++++----
 1 file changed, 25 insertions(+), 5 deletions(-)

diff --git a/implementation/service_discovery/src/service_discovery_impl.cpp b/implementation/service_discovery/src/service_discovery_impl.cpp
index 3eb629da3..0aa69ef45 100644
--- a/implementation/service_discovery/src/service_discovery_impl.cpp
+++ b/implementation/service_discovery/src/service_discovery_impl.cpp
@@ -114,11 +114,31 @@ service_discovery_impl::init() {
         initial_delay_max = tmp;
     }
 
-    std::random_device r;
-    std::mt19937 e(r());
-    std::uniform_int_distribution<std::uint32_t> distribution(
-            initial_delay_min, initial_delay_max);
-    initial_delay_ = std::chrono::milliseconds(distribution(e));
+    try {
+        std::random_device r;
+        std::mt19937 e(r());
+        std::uniform_int_distribution<std::uint32_t> distribution(
+                initial_delay_min, initial_delay_max);
+        initial_delay_ = std::chrono::milliseconds(distribution(e));
+    } catch (std::exception const& e) {
+        VSOMEIP_ERROR << "Failed to generate random initial delay: " << e.what();
+
+        // Fallback to the Mersenne Twister engine
+        auto const seed = static_cast<std::mt19937::result_type>(
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::high_resolution_clock::now().time_since_epoch())
+                .count());
+
+        std::mt19937 mtwister{seed};
+
+        // Interpolate between initial_delay bounds
+        initial_delay_ = std::chrono::milliseconds(
+            initial_delay_min +
+            (static_cast<std::int64_t>(mtwister()) *
+             static_cast<std::int64_t>(initial_delay_max - initial_delay_min) /
+             static_cast<std::int64_t>(std::mt19937::max() -
+                                       std::mt19937::min())));
+    }
 
 
     repetitions_base_delay_ = std::chrono::milliseconds(
